src/mapgen/cavegen.cpp

```diff
@@ -247,22 +263,22 @@ bool CavernsNoise::generateCaverns(MMVManip *vm, v3s16 nmin, v3s16 nmax)
 		for (s16 y = nmax.Y; y >= nmin.Y - 1; y--,
 				index3d -= m_ystride,
 				VoxelArea::add_y(em, vi, -1),
-				cavern_amp_index++) {
+				cavern_offset_index++) {
 			content_t c = vm->m_data[vi].getContent();
-			float n_absamp_cavern = std::fabs(noise_cavern->result[index3d]) *
-				cavern_amp[cavern_amp_index];
+			float n_absoff_cavern = std::fabs(noise_cavern->result[index3d]) +
+				cavern_offset[cavern_offset_index];
```

Comment:
This is the key point in improving the cavern tapering, instead of multiplying the noise by a '0 to 1 amplitude',
a '-3 to 0 offset' is added to the noise.